src/hash.cr

```diff
@@ -1690,8 +1690,18 @@ class Hash(K, V)
   # hash.transform_keys { |key, value| key.to_s * value } # => {"a" => 1, "bb" => 2, "ccc" => 3}
   # ```
   def transform_keys(& : K, V -> K2) : Hash(K2, V) forall K2
-    each_with_object({} of K2 => V) do |(key, value), memo|
-      memo[yield(key, value)] = value
+    if size <= 4
+      # Building up an empty hash is faster than a pre-allocated one as long as
+      # the size is below 5.
+      each_with_object({} of K2 => V) do |(key, value), memo|
+        memo[yield(key, value)] = value
+      end
+    else
+      hash = Hash(K2, V).new(initial_capacity: size)
+      each do |key, value|
+        hash[yield(key, value)] = value
+      end
+      hash
```

Comment:
Could we merge the branches? Like:

```crystal
copy = size > 4 ? Hash(K2, V).new(capacity) : Hash(K2, V).new
each_with_object(copy) { ... }
```

Maybe 4 could be a constant instead of a magic number? For example `Hash::INITIAL_CAPACITY` or `Hash::MINIMAL_CAPACITY`.

Reply:
It can be even simpler: `Hash(K2, V).new(size < 4 ? 0 : size)`